lib/game_factory.rb

```diff
@@ -0,0 +1,47 @@
+class GameFactory
+  attr_reader :player_one, :player_two
+
+  def initialize(board, display, rules)
+    @board = board
+    @display = display
+    @rules = rules
+  end
+
+  def create_game
+    create_players
+    game = Game.new(@display, @rules, @player_one, @player_two)
+    game.start
+  end
+
+  private
+
+  def create_players
+    player_one_mark = create_mark
+    player_two_mark = create_mark(player_one_mark)
+    @player_one = HumanPlayer.new(player_one_mark, @board, @display)
+    @player_two = HumanPlayer.new(player_two_mark, @board, @display)
+  end
+
+  def create_mark(other_mark = nil)
+    @display.setup_mark_message
+    potential_mark = @display.retrieve_user_input
+    mark = check_game_mark(potential_mark, other_mark)
+    @mark = mark
+  end
+
+  def check_game_mark(potential_mark, other_mark)
+    until valid_game_piece?(potential_mark, other_mark)
+      @display.invalid_mark_message
+      potential_mark = @display.retrieve_user_input
+    end
+    potential_mark
+  end
+
+  def valid_game_piece?(potential_mark, other_mark)
+    if potential_mark.length > 1 || potential_mark == ' ' || potential_mark == '' || potential_mark == other_mark
+      false
+    else
+      true
```

Comment:
First, you have `potential_mark == ''` twice.
You can save a few lines by writing the above as
```ruby
def valid_game_piece?(potential_mark, other_mark)
  true unless potential_mark.length > 1 || potential_mark == ' ' || potential_mark == other_mark
end
```

But here's a tip for functions that check a condition and then return `true` or `false` - you don't have to explicitly return `true`/`false`

I usually try to avoid explicitly returning true/false when the answer is completely based on a condition, but that's a little harder to do here since your conditions are written for the `false` case. 

Given the boolean expressions that you've written, you can have a one line function if you change your function name to be `invalid` game piece, and if you want to use this, you'll need to change your `check_game_mark` method.
```ruby
def invalid_game_piece?(potential_mark, other_mark)
  potential_mark.length > 1 || potential_mark == ' ' || potential_mark == other_mark
end
```

Or, if you want this function to continue to be called `valid` (and not change how your `check_game_mark` method works), we need to think of what conditions are necessary for the piece to be **valid** (since currently you've written the conditions for the invalid case)

I don't remember if we said that game marks could be more than one character, but assuming they need to be one character:
```ruby
def valid_game_piece?(potential_mark, other_mark)
  potential_mark.length == 1 && potential_mark != ' '  && potential_mark != other_mark
end
```
Now that I've written this out, I realize that `potential_mark.length == 1 && potential_mark != ' '` are actually redundant - if we restrict the length to 1, then the mark `''` won't be valid, and we could remove that condition, leaving us with 

```ruby
def valid_game_piece?(potential_mark, other_mark)
  potential_mark.length == 1  && potential_mark != other_mark
end
```

😄

Reply:
This was a very good code walkthrough!